deltacat/exceptions.py

```diff
@@ -1,14 +1,145 @@
-class RetryableError(Exception):
-    pass
+from __future__ import annotations
+from enum import Enum
 
 
-class NonRetryableError(Exception):
-    pass
+def _pack_all_args(exception_cls, args=None, kwargs=None):
+    # This is helpful for reducing Exceptions that only accept kwargs as
+    # only positional arguments can be provided for __reduce__
+    if args is None:
+        args = ()
+    if kwargs is None:
+        kwargs = {}
+    return exception_cls(*args, **kwargs)
 
 
-class ConcurrentModificationError(Exception):
-    pass
+class DeltaCatError(Exception):
+    msg = "An unexpected DeltaCat Error occurred."
+    error_code = "10000"
+    is_retryable = False
 
+    def __init__(self, **kwargs):
+        self.error_info = self.msg.format(
+            **kwargs, error_code=self.error_code, is_retryable=self.is_retryable
+        )
+        Exception.__init__(self, self.error_info)
 
-class ValidationError(NonRetryableError):
-    pass
+    def __reduce__(self):
+        return _pack_all_args(
+            (
+                self.__class__,
+                (self.error_code, self.is_retryable),
+                self.kwargs,
+            )
+        )
+
+
+class DeltaCatErrorMapping(Exception, Enum):
+
+    # Dependency Error code from 10100 to 10199
+    GeneralDependencyError = "10100"
```

Comment:
Why can't this exception/enum contain error code, whether its retryable and error message altogether?

Reply:
The primary purpose for this enum is to show which sub category this error is belonged to like it belongs to dependency error. I think it looks clearer with long error message goes under each class definition?

Reply:
Ok. enum keys must be in caps.